src/vs/workbench/contrib/chat/browser/chatContentParts/chatPrepareToolInvocationPart.ts

```diff
@@ -0,0 +1,117 @@
+
+		this.element = context.element;
+
+		// Hide when following content contains parts that are not prepareToolInvocation or progressMessage
+		// This is similar to how progressMessage parts hide when other content arrives
+		const followingContent = context.content.slice(context.contentIndex + 1);
+		this.isHiddenByFollowingContent = followingContent.some(part => part.kind !== 'prepareToolInvocation' && part.kind !== 'progressMessage');
+
+		// Asynchronously call handleToolStream and update the message
+		if (!this.isHiddenByFollowingContent) {
+			this.fetchAndUpdateMessage(prepareToolInvocation);
+		}
+	}
+
+	private static createInitialProgressMessage(prepareToolInvocation: IChatPrepareToolInvocationPart, languageModelToolsService: ILanguageModelToolsService): IChatProgressMessage {
+		const toolData = languageModelToolsService.getTool(prepareToolInvocation.toolName);
+		const displayName = toolData?.displayName ?? prepareToolInvocation.toolName;
+		return {
+			kind: 'progressMessage',
+			content: new MarkdownString(localize('invokingTool', "Invoking tool: {0}", displayName))
+		};
+	}
+
+	private async fetchAndUpdateMessage(prepareToolInvocation: IChatPrepareToolInvocationPart): Promise<void> {
+		const toolData = this.languageModelToolsService.getTool(prepareToolInvocation.toolName);
+		if (!toolData) {
+			return;
+		}
+
+		const streamContext: IToolInvocationStreamContext = {
+			toolCallId: prepareToolInvocation.toolCallId,
+			rawInput: prepareToolInvocation.streamData?.partialInput,
+			chatRequestId: isResponseVM(this.element) ? this.element.requestId : undefined,
+			chatSessionId: this.element.sessionId
+		};
+
+		try {
+			const streamResult = await this.languageModelToolsService.handleToolStream(toolData.id, streamContext, CancellationToken.None);
+			if (streamResult?.invocationMessage) {
+				const progressContent = typeof streamResult.invocationMessage === 'string'
+					? new MarkdownString(streamResult.invocationMessage)
+					: new MarkdownString(streamResult.invocationMessage.value, { isTrusted: streamResult.invocationMessage.isTrusted, supportThemeIcons: streamResult.invocationMessage.supportThemeIcons, supportHtml: streamResult.invocationMessage.supportHtml });
+				this.updateMessage(progressContent);
+			}
+		} catch (error) {
+			this.logService.warn(`ChatPrepareToolInvocationPart: Error calling handleToolStream for tool ${toolData.id}`, error);
+		}
+	}
+
+	override hasSameContent(other: IChatRendererContent, followingContent: IChatRendererContent[], element: ChatTreeItem): boolean {
+		if (other.kind !== 'prepareToolInvocation') {
+			return false;
+		}
+
+		// If following content contains parts that are not prepareToolInvocation or progressMessage,
+		// we need to re-render to hide this part (similar to progressMessage behavior)
+		const shouldBeHidden = followingContent.some(part => part.kind !== 'prepareToolInvocation' && part.kind !== 'progressMessage');
```

Comment:
I don't think we should follow this pattern of hiding things. Now we have a toolcallid so we can associate the incomplete call with the real call, it seems to me like the incoplete call should just sort of resolve into the real call widget.